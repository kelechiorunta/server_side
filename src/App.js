import React, { useCallback, useState } from 'react';
import RefComponent from './components/RefComponent.jsx';
import { useRef } from 'react';
import Tooltip from './components/Tooltip.jsx';

export const addThreeToApp = (num) => {
  return num + 3;
};

function App() {
  const [surname, setSurname] = useState('');
  const inputRef = useRef(null);
  const divRef = useRef(null);
  const [targetRect, setTargetRect] = useState(null);
  const handleSelect = (e) => {
    setSurname('Kelechi');
    // setSurname(e.target.innerText);
  };

  // Callback Ref meant to update divRef.current
  const setDivRef = useCallback((node) => {
    if (node) {
      divRef.current = node;
    }
  }, []);

  const handleFocus = () => {
    inputRef.current.focus();
    inputRef.current.scrollDown();
  };

  const handlePointer = (e) => {
    // if (divRef && divRef.current) {
    divRef.current = e.target;
    const { left, right, top, bottom } = e.target.getBoundingClientRect();
    setTargetRect({
      left: left,
      top: top,
      right: right,
      bottom: bottom
    });
    // }
  };

  const handlePointerLeave = () => {
    // const { left, right, top, bottom } = ref.current.getBoundingClientRect();
    setTargetRect(null);
  };

  return (
    <div className="App">
      <header className="App-header">
        {/* <img src={logo} className="App-logo" alt="logo" /> */}
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        Learn React
        {surname && <h1>{surname}</h1>}
        <RefComponent
          placeholder={'Enter Your Name'}
          type={'text'}
          style={{ borderRadius: 5, border: '2px solid black' }}
          ref={inputRef}
        />
        <div
          style={{ cursor: 'pointer' }}
          ref={setDivRef}
          onPointerEnter={handlePointer}
          onPointerLeave={() => handlePointerLeave()}
        >
          Hello Maggie
        </div>
        {divRef && targetRect && (
          <Tooltip
            ref={divRef}
            targetRect={targetRect}
            content={<div>{divRef.current.textContent}</div>}
          />
        )}
        <button
          style={{ cursor: 'pointer', marginTop: 50 }}
          ref={setDivRef}
          onPointerEnter={handlePointer}
          onPointerLeave={() => handlePointerLeave()}
          onClick={handleFocus}
        >
          Focus
        </button>
        <button
          style={{ cursor: 'pointer' }}
          ref={setDivRef}
          onPointerEnter={handlePointer}
          onPointerLeave={() => handlePointerLeave()}
          onClick={handleSelect}
        >
          Show Surname
        </button>
      </header>
    </div>
  );
}

export default App;
