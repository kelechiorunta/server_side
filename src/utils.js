export const addTwoToAge = (age) => {
  if (isNaN(age) || !age) return;

  return age + 2;
};
