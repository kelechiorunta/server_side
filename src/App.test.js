import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { createRoot } from 'react-dom/client';
import App from './App';

// ✅ Import the function normally if you're not mocking it in this test
import * as utils from './utils';
import { addThreeToApp } from './App';

jest.mock('./utils', () => ({
  addTwoToAge: jest.fn(() => 4) // ✅ Mocked return value
}));

describe('App', () => {
  it('renders App component correctly', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('mocks addTwoToAge correctly', () => {
    const result = utils.addTwoToAge(2);
    expect(utils.addTwoToAge).toHaveBeenCalledWith(2);
    expect(result).toBe(4);
  });

  it('mocks the addition 3 correctly', () => {
    const mocked = jest.fn().mockImplementation((n) => ({ add: n + 1 }));
    const mockedfn = mocked.mockName;
    const result = mocked(1).add;

    console.log(mockedfn);

    expect(mocked).toHaveBeenCalledWith(1);
    expect(result).toBe(2);
  });

  it('gets the button click handler', async () => {
    render(<App />);
    const testbutton = screen.getByText(/Show Surname/i);

    fireEvent.click(testbutton);

    const result = await screen.findByText(/Kelechi/i);
    expect(result).toBeVisible();
  });
});

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });
