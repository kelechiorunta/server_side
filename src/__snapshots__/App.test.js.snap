// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component correctly 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <p>
        Edit 
        <code>
          src/App.js
        </code>
         and save to reload.
      </p>
      <a
        class="App-link"
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
      <button>
        Show Surname
      </button>
    </header>
  </div>
</div>
`;
